lib/puma/client.rb

```diff
@@ -273,6 +273,17 @@ def peerip
       @peerip ||= @io.peeraddr.last
     end
 
+    # Returns true if the connection is idle and may be closed.
+    #
+    # From RFC 2616 section 8.1.4:
+    # Servers SHOULD always respond to at least one request per connection,
+    # if at all possible. Servers SHOULD NOT close a connection in the
+    # middle of transmitting a response, unless a network or client failure
+    # is suspected.
+    def idle?
+      @requests_served > 0 && !in_data_phase
```

Comment:
Why would we not be idle if `@requests_served` is 0?

Reply:
I interpreted the RFC statement "Servers SHOULD always respond to at least one request per connection, if at all possible" to mean that the server shouldn't consider unused (not served at least one request) connections to be 'idle', so shouldn't force-close them during a graceful shutdown. This makes sense to me- if a client opens a connection it usually makes at least one request pretty soon after.

Maybe `#idle?` isn't the clearest term for this method? I used it here because it the same [RFC section](https://tools.ietf.org/html/rfc2616#section-8.1.4) uses it to refer to a connection 'the server has decided to close':

>    A client, server, or proxy MAY close the transport connection at any
   time. For example, a client might have started to send a new request
   at the same time that the server has decided to close the "idle"
   connection. From the server's point of view, the connection is being
   closed while it was idle, but from the client's point of view, a
   request is in progress.

I also considered `#can_close?`, `#should_close?` or `#closeable?` if any of those seem more fitting.

Reply:
Riiight, so if we haven't served any requests, we can just shut down. Got it.

Any of the `close` methods you proposed I think would be more fitting.

Reply:
I don't think `@requests_served` is a good variable name here. It is (from what I understand) a count of the number of requests that have been received from the client, and not the number of requests served from the server. I think a better name might be `@client_request_count`.

Given that tidbit, I don't think the logic here is entirely correct. I do want to take a step back from the specific implementation though and make sure we're on the same page at a high level.

Reply:
Given that this on client.rb, I'm OK with the variable name.